STM/src/test/java/org/jboss/stm/TaxonomyNestedTest.java

```diff
@@ -0,0 +1,407 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.stm;
+
+import com.arjuna.ats.arjuna.AtomicAction;
+import com.arjuna.ats.arjuna.coordinator.ActionStatus;
+import junit.framework.TestCase;
+import org.jboss.stm.annotations.Nested;
+import org.jboss.stm.annotations.ReadLock;
+import org.jboss.stm.annotations.Transactional;
+import org.jboss.stm.annotations.WriteLock;
+import org.jboss.stm.internal.RecoverableContainer;
+import org.junit.BeforeClass;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+// these tests are discussed in the article https://jbossts.blogspot.com/2018/09/tips-on-how-to-evaluate-stm.html
+// and relate to the behaviour of nested transactions
+public class TaxonomyNestedTest extends TestCase {
+    @Transactional // define some transactional memory
+    @Nested // with nesting semantics
+
+    public interface AtomicInt {
+        int get() throws Exception;
+        void set(int value) throws Exception;
+    }
+
+    public class AtomicIntImpl implements AtomicInt {
+        private int state;
+
+        @ReadLock
+        public int get() throws Exception {
+            return state;
+        }
+
+        @WriteLock
+        public void set(int value) throws Exception {
+            state = value;
+        }
+    }
+
+    // define some types for testing running nested transactions in a thread that is
+    // different from the thread used to run the top level transaction
+    private enum MSG_TYPE {
+        EXIT, // indication for the thread to exit
+        GET, // the thread should perform a transactional read
+        SET // the thread should perform a transactional write
+    };
+
+    // define a message type for the parent and child threads to communicate their behaviour
+    private static class Message {
+        MSG_TYPE what; // the behaviour that the thread should perform
+        Integer value; // the value to write if the behaviour is write
+        int result; // the value read if the behaviour is read
+        Throwable throwable; // the throwable if the read or write failed
+
+        private Message(MSG_TYPE what, Integer value, int result, Throwable throwable) {
+            this.what = what;
+            this.value = value;
+            this.result = result;
+            this.throwable = throwable;
+        }
+
+        Message(MSG_TYPE what, Integer value) {
+            this(what, value, -1, null);
+        }
+    }
+
+    enum NESTED_TYPE {TOP_LEVEL, COMMIT, ABORT};
+
+    // define a task for performing nested reads and writes to transactional memory
+    private class ThreadedReaderWriter extends Thread {
```

Comment:
I assume this inner class is quite complex. I could be good to consider it for being a "standalone" class.

Reply:
It is not particularly complex and rightly belongs to the class it resides in. If I ever need to reuse its functionality (in non nested scenarios) in later blogs in this series about STM then it would make sense to refactor it into a separate class.

Reply:
For example, the next article in this series will cover concurrency (thread interaction, pessimistic, optimistic, when does the STM implementation detects failures, ... etc) and I would imagine this inner class could be extended to support that at which point it would make sense to refactor it into its own class file.